### PR TITLE
refactor(wire-protocol): flatten wire protocol handler hierarchy

### PR DESCRIPTION
**Pull request by mbroadst**
See the ticket for more details, but this generally removes the
class-based module design for wire protocol handling, removing
needless allocations, and focusing more on the meat of the methods
hung off those classes.

NODE-1835

### CODE REVIEW

File: lib/wireprotocol/get_more.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        ? Long.fromNumber(response.documents[0].cursor.id)
+        : response.documents[0].cursor.id;
+
+    cursorState.documents = response.documents[0].cursor.nextBatch;
+    cursorState.cursorId = cursorId;
+
+    callback(null, response.documents[0], response.connection);
+  }
+
+  if (wireVersion < 4) {
```
Comment (daprahamian) [inline lib/wireprotocol/get_more.js]: Side note: I'd like to unify all of our wire-version checks behind a layer of encapsulation. At the bare minimum, constants for wireVersions.

File: lib/wireprotocol/write_command.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+const MongoError = require('../error').MongoError;
+const collectionNamespace = require('./shared').collectionNamespace;
+const command = require('./command');
+
+function writeCommand(server, type, opsField, ns, ops, options, callback) {
+  if (ops.length === 0) throw new MongoError('insert must contain at least one document');
```
Comment (daprahamian) [inline lib/wireprotocol/write_command.js]: perhaps error message should be 
```
`${type} must contain at least one document`
```

File: lib/wireprotocol/write_command.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
```
Comment (daprahamian) [inline lib/wireprotocol/write_command.js]: What are these two lines supposed to do?

File: lib/wireprotocol/query.js [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+  if (cmd.min) findCmd.min = cmd.min;
+  if (cmd.max) findCmd.max = cmd.max;
+  findCmd.returnKey = cmd.returnKey ? cmd.returnKey : false;
+  findCmd.showRecordId = cmd.showDiskLoc ? cmd.showDiskLoc : false;
+  if (cmd.snapshot) findCmd.snapshot = cmd.snapshot;
+  if (cmd.tailable) findCmd.tailable = cmd.tailable;
+  if (cmd.oplogReplay) findCmd.oplogReplay = cmd.oplogReplay;
+  if (cmd.noCursorTimeout) findCmd.noCursorTimeout = cmd.noCursorTimeout;
+  if (cmd.awaitData) findCmd.awaitData = cmd.awaitData;
+  if (cmd.awaitdata) findCmd.awaitData = cmd.awaitdata;
```
Comment (kvwalker) [inline lib/wireprotocol/query.js]: There are 2 of this line.

File: lib/wireprotocol/query.js [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+function prepareLegacyFindQuery(server, ns, cmd, cursorState, options) {
+  options = options || {};
+  const bson = server.s.bson;
+  const readPreference = getReadPreference(cmd, options);
+  cursorState.batchSize = cmd.batchSize || cursorState.batchSize;
+
+  let numberToReturn = 0;
+  if (cursorState.limit === 0) {
```
Comment (kvwalker) [inline lib/wireprotocol/query.js]: Could we simplify this check to be the following?
```
if (cursorState.limit < 0 ||
    (cursorState.limit !== 0 && (cursorState.limit < cursorState.batchSize)) ||
    (cursorState.limit > 0 && cursorState.batchSize === 0)
) { 
  ...
} else {
  numberToReturn = cursorState.batchSize;
}
```

File: lib/wireprotocol/get_more.js (hunk 1 above)
Comment (mbroadst) [reply lib/wireprotocol/get_more.js]: totally agree

File: lib/wireprotocol/write_command.js (hunk 3 above)
Comment (mbroadst) [reply lib/wireprotocol/write_command.js]: looks like merge pasta 😄 fixed

File: lib/wireprotocol/query.js (hunk 4 above)
Comment (mbroadst) [reply lib/wireprotocol/query.js]: it's actually checking for an all lowercase version, unclear whether that is still an issue in the future but I'm hesitant to change it

Comment (kvwalker) [reply lib/wireprotocol/query.js]: Oh, good point!

File: lib/wireprotocol/query.js (hunk 5 above)
Comment (mbroadst) [reply lib/wireprotocol/query.js]: yep!

File: lib/wireprotocol/write_command.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
+'use strict';
+
+const MongoError = require('../error').MongoError;
+const collectionNamespace = require('./shared').collectionNamespace;
+const command = require('./command');
+
+function writeCommand(server, type, opsField, ns, ops, options, callback) {
+  if (ops.length === 0) throw new MongoError(`${type} must contain at least one document`);
+  if (typeof options === 'function') (callback = options), (options = {});
```
Comment (daprahamian) [inline lib/wireprotocol/write_command.js]: Please put these on two lines.

**Comment by mbroadst** [comment]
@kvwalker @daprahamian all issues should have been resolved, PTAL.

**Pull request merged**